tests/functional/test_main.py

```diff
@@ -0,0 +1,229 @@
+from http import HTTPStatus
+from flask_login import login_user, current_user
+from flask import session
+from pytest_mock import mocker
+from unittest import mock
+from app.search import LastFMSearch, YTSearch
+
+
+def test_root(client, app, test_db):
+
+    response = client.get("/")
+    assert response.status_code == HTTPStatus.OK
+
+
+def test_login_get(client, app):
+
+    response = client.get("login")
+    assert response.status_code == HTTPStatus.OK
+
+
+def test_login_post(client, app, persisted_user):
+
+    response = client.post(
+        "login",
+        data={"username": persisted_user.username, "password": "unhashed-pw"},
+        follow_redirects=True,
+    )
+    assert b"Welcome back" in response.data
+
+
+def test_login_post_bad_pw(client, app, persisted_user):
+
+    response = client.post(
+        "login",
+        data={"username": persisted_user.username, "password": "wrong_password"},
+        follow_redirects=True,
+    )
+    assert b"Invalid credentials" in response.data
+
+
+def test_login_post_bad_username(client, app, persisted_user):
+
+    response = client.post(
+        "login",
+        data={"username": "badusername", "password": "unhashed-pw"},
+        follow_redirects=True,
+    )
+    assert b"Username not found" in response.data
+
+
+def test_signup_get(app):
+
+    with app.test_client() as test_client:
+        response = test_client.get("/signup")
+        assert response.status_code == HTTPStatus.OK
+        assert b"Sign up for Check Out This Song!" in response.data
+
+
+def test_signup_post(client, app, test_db):
+
+    response = client.post(
+        "/signup",
+        data={
+            "username": "testSignupUser",
+            "password": "test_signup_password",
+            "email": "test_email@email.com",
+            "avatar": None,
+            "bio": "",
+        },
+        follow_redirects=True,
+    )
+    print(response.data)
+    assert b"Welcome to Check Out This Song" in response.data
+
+
+def test_signup_post_illegal_username(client, app, test_db):
+
+    response = client.post(
+        "/signup",
+        data={
+            "username": "test_signup_user",
+            "password": "test_signup_password",
+            "email": "test_email@email.com",
+            "avatar": None,
+            "bio": "",
+        },
+        follow_redirects=True,
+    )
+    assert b'<p class="help is-danger">' in response.data
```

Comment:
Could make this more clear by doing something like this:

```
def test_signup_post_illegal_username(client, app, test_db, valid_user):
     valid_user.username = "this username is illegal"
     response = client.post(
         "/signup",
         data=build_request_from_user(valid_user),
         follow_redirects=True,
     )
     assert b'<p class="help is-danger">' in response.data
```

https://www.destroyallsoftware.com/screencasts/catalog/test-data-factories